### PR TITLE
Gate proposal diff on expiry and structure to match apply (#1376)

### LINKED ISSUE #1376: Preview/apply asymmetries remaining after #1370: expiry not checked at diff time; zero-op proposals diverge (200/404 vs 400)
## Context

PR #1374 (issue #1370) restored `preview == apply` for reorder positions, create-card id collisions, and structure limits. Two adversarial review lenses plus a bot review surfaced two REMAINING asymmetries in the same trust class. Both are **pre-existing** (not introduced by #1374) and were adjudicated out of scope there; tracked here per the zero-skip review policy.

## Asymmetries

1. **Proposal expiry is not checked at diff time.** `AutomationProposalService.GetProposalDiffAsync` never checks `ExpiresAt`, while apply's `AutomationPolicyEngine.ValidatePolicy` rejects expired proposals ("Proposal has expired"). An expired proposal previews a clean diff, then Apply fails after approval — the same preview-then-apply-fails trust violation class #1370 fixed.

2. **Zero-operation proposals diverge between diff and apply.** The diff path's zero-op branch returns:
   - **200** with the cached `DiffPreview` when one is stored (flagged independently by `chatgpt-codex-connector` on PR #1374: https://github.com/Chris0Jeky/Taskdeck/pull/1374#discussion_r3599630223), or
   - **404** "Diff preview not available for this proposal" when none is stored,

   while apply's structure validation returns **400** "Proposal must contain at least one operation". A reviewer can preview a proposal that apply will always reject, and the two failure shapes disagree.

## Suggested direction

Run the same gates the executor runs (expiry + structure) before returning any diff — including the cached-`DiffPreview` fast paths — returning the identical `ValidationError` shapes. Mirror how #1374 wired `ProposalOperationStructureValidator` into the original-proposal diff path.

## Provenance

- PR #1374 consolidated review: https://github.com/Chris0Jeky/Taskdeck/pull/1374#issuecomment-4997926127
- Surfaced by the 2026-07-17 adversarial review adjudication over #1370/#1374.

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Two pre-existing preview/apply asymmetries remained in the same trust class #1374 (issue #1370) started closing: preview must reject exactly what apply rejects, or a reviewer approves a diff that then fails after approval. `AutomationProposalService.GetProposalDiffAsync` did not run two gates the executor runs via `AutomationPolicyEngine.ValidatePolicy`.

This wires both the **structure** gate and the **expiry** gate into every diff path — the revised-payload path, the original-operations path, and the cached-`DiffPreview` fast paths — in `ValidatePolicy`'s order (structure, then expiry), returning the identical `ValidationError` shapes apply returns.

Closes #1376

### 1. Proposal expiry was not checked at diff time
`GetProposalDiffAsync` never checked `ExpiresAt`, while apply's `ValidatePolicy` rejects expired proposals. An expired proposal previewed a clean diff (or returned its cached `DiffPreview`), then apply failed after approval.

### 2. Zero-operation proposals diverged between diff and apply
The diff path's zero-op branch returned the cached `DiffPreview` (200) when one was stored, or 404 "Diff preview not available for this proposal" when none was — while apply's structure validation returns 400 "Proposal must contain at least one operation". A reviewer could preview a proposal apply always rejects, and the failure shapes disagreed.

## Before / after HTTP codes + error shapes

`ErrorCodes.ValidationError` maps to HTTP 400; `ErrorCodes.NotFound` maps to 404 (`ResultExtensions`).

| Defect | Diff before | Diff after | Apply (unchanged) |
|---|---|---|---|
| Expired, well-formed, non-empty (no cache) | 200 rendered diff | **400** `ValidationError` "Proposal has expired" | 400 `ValidationError` "Proposal has expired" |
| Expired, with cached `DiffPreview` | 200 cached preview | **400** `ValidationError` "Proposal has expired" | 400 `ValidationError` "Proposal has expired" |
| Expired, with saved revision | 200 revised diff | **400** `ValidationError` "Proposal has expired" | 400 `ValidationError` "Proposal has expired" |
| Zero-op, no cached preview | 404 `NotFound` "Diff preview not available for this proposal" | **400** `ValidationError` "Proposal must contain at least one operation" | 400 `ValidationError` "Proposal must contain at least one operation" |
| Zero-op, with cached `DiffPreview` | 200 cached preview | **400** `ValidationError` "Proposal must contain at least one operation" | 400 `ValidationError` "Proposal must contain at least one operation" |

Ordering matches `ValidatePolicy`: structure runs before expiry, so a proposal that is both empty and expired reports the empty error on both boundaries.

## Back-compat

A non-expired, well-formed proposal's diff is byte-identical to today — the new gates pass cleanly and the cached-`DiffPreview` fast path still returns the stored preview unchanged (covered by a regression test).

## Implementation

- Added a `ValidateProposalNotExpired` helper mirroring `ValidatePolicy`'s expiry expression (`DateTime.UtcNow > proposal.ExpiresAt` → `ValidationError` / "Proposal has expired"). No apply-side behavior changed.
- Original path: expiry gate added after the existing structure gate (from #1374), ahead of the contract check and cached-preview return.
- Zero-op branch removed — flow now falls through to the structure gate, which returns the apply-shaped 400 regardless of a stored `DiffPreview`.
- Revised path: structure gate + expiry gate added after payload parse, before the contract check (apply re-runs structure+expiry on the materialized revision).

## Verification

Backend build: `dotnet build backend/Taskdeck.sln -c Release` — 0 errors (pre-existing warnings only).

Targeted:
- `AutomationProposalServiceTests` — 57 passed (includes 5 new/retargeted diff-parity tests: expiry parity vs `ValidatePolicy`, expiry with cached preview, expiry on revised path, zero-op with and without stored preview, plus the non-empty cached-preview back-compat regression).
- Parity/policy/executor/validator classes (`ProposalOperationStructureValidatorTests`, `ProposalOperationContractValidatorTests`, `AutomationPolicyEngineTests`, `AutomationPolicyEngineEdgeCaseTests`, `AutomationProposalServiceEdgeCaseTests`, `AutomationExecutorServiceTests`, `ColumnServiceTests`) — 115 passed.
- `McpToolsTests` (Api.Tests, exercises `GetProposalDiffAsync` end-to-end) — 29 passed.

Project-level: `Taskdeck.Application.Tests` — **3488 passed, 0 failed, 0 skipped**.

Full-solution suite intentionally not run here (coordinator owns it serially at gate time).

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Adversarial Code Review

Independent self-review plus triage of bot findings on this PR.

### CRITICAL
- None.

### HIGH
- None. Verified the revised-path expiry parity is exact: apply's executor materializes the effective proposal via `proposal with { Operations = revisedOperations }` (`AutomationExecutorService.MaterializeEffectiveProposalAsync`), which preserves `ExpiresAt`, and `ValidatePolicy` runs structure-then-expiry on it. The diff path checks the same entity `ExpiresAt` in the same order, so preview and apply reject identically on every path (revised, original, cached fast paths).

### MEDIUM
- `ValidateProposalNotExpired` duplicated the domain expiry expression (`DateTime.UtcNow > proposal.ExpiresAt`) instead of reusing `AutomationProposal.IsExpired`. Matches gemini-code-assist's finding. Fixed — now delegates to the domain property (one source of truth); semantics and the `ValidationError` / "Proposal has expired" shape are unchanged.

### LOW
- The `SetExpiresAt` reflection helper is duplicated between `AutomationProposalServiceTests` and `AutomationProposalServiceEdgeCaseTests`. Left as-is: this mirrors the established test convention (the edge-case file already carries its own copy), and extracting a shared test utility would touch a second test file outside this PR's seam, risking collision with concurrent work. Not merge-blocking; noted for a future test-util consolidation.

### Bot Comments Addressed
- gemini-code-assist (MEDIUM, `AutomationProposalService.cs:488`): reuse domain `IsExpired`. Fixed in `9f6eee1b`.
- copilot-pull-request-reviewer: could not review (reviewer quota limit) — no findings to address.

### Summary
0 CRITICAL, 0 HIGH, 1 MEDIUM (fixed), 1 LOW (accepted with rationale). No merge-blocking issues remain. Back-compat verified: a non-expired, well-formed proposal's diff (including the cached-preview fast path) is byte-identical to before.

**Comment by Chris0Jeky** [comment]
## Adversarial Review — Fixes Applied

| Finding | Severity | Fix Commit | Verified |
|---------|----------|-----------|----------|
| Reuse domain `IsExpired` instead of duplicating the expiry expression (gemini-code-assist) | MEDIUM | `9f6eee1b` | `AutomationProposalServiceTests` 57 passed; Release build clean |
| `SetExpiresAt` test-helper duplication across test classes | LOW | — (accepted) | Cross-file dedup out of scope for this seam; matches existing convention |
| copilot-pull-request-reviewer quota limit — no review produced | n/a | — | No findings |

All actionable findings addressed. This PR touches no migrations/EF model. No CRITICAL/HIGH findings.

**Comment by Chris0Jeky** [comment]
## Adversarial Review — Round 2 (consolidated, two independent lenses)

Coordinator-adjudicated round 2: one **parity lens** (backend trust-surface correctness) and one **consumer lens** (every caller of the changed contract). Verdicts below.

### Parity lens — core claims CONFIRMED
- **Gate order** matches Apply exactly: structure then expiry, mirroring `AutomationPolicyEngine.ValidatePolicy`.
- **Error-shape identity** holds structurally, not just by string copy: the structure gate is the shared `ProposalOperationStructureValidator` (single source of truth), and both boundaries map through the same `ToErrorActionResult` (`ValidationError` → 400).
- **Expiry semantics are tick-identical**: both sides evaluate the domain expression (`DateTime.UtcNow > ExpiresAt`; preview now via `AutomationProposal.IsExpired` after the round-1 fix).
- **Revised-path preservation**: `MaterializeEffectiveProposalAsync` uses `proposal with { Operations = ... }`, preserving `ExpiresAt`, so the revised diff path checks the same instant Apply checks.
- **Cache fast-path intact**: a non-expired, well-formed proposal's cached `DiffPreview` return is byte-identical to before (regression-tested).

### Consumer lens — confirmed-safe coverage
- **Controller mapping** (`AutomationProposalsController` `/diff`): pass-through `ToErrorActionResult`; no code-specific branching to break.
- **MCP resource reads the stored field**, not the live diff endpoint — unaffected.
- **No other backend callers** of `GetProposalDiffAsync` beyond the controller and tests (`McpToolsTests` exercises non-expired, well-formed proposals; 29 passed).
- **No status-code branching in frontend diff handlers** — both shells funnel all errors through `getErrorDisplay` → toast, so nothing crashes; the findings below are presentation/coverage quality, not breakage.

### Findings and dispositions

| # | Severity | Finding | Disposition |
|---|----------|---------|-------------|
| F1 | LOW | `PaperReviewView.vue` carried now-stale comments (~670-676, ~707-710) asserting the no-op diff path returns 404; after this PR it returns 400 "Proposal must contain at least one operation" (404 now only means proposal-not-found) | **Fixed in-PR** (comment-only; no behavior change) |
| A1 | MEDIUM (UX/test debt, out of PR scope) | Legacy ReviewView (`useReviewActions.ts:114-130`) becomes inspect-blind for EXPIRED proposals with operations (400 → toast + cleared pane; no inline fallback like Paper's stored `diffPreview`); product call needed on expired-proposal inspectability | **Tracked: #1397** |
| A2 | LOW (degenerate state, out of PR scope) | Zero-op proposal WITH cached `diffPreview` bypasses PaperReviewView's no-op guard (~677-689 checks `!p.diffPreview` only) and now toasts where it previously rendered the stale cached preview | **Tracked: #1397** |
| A3 | MEDIUM (test gap, out of PR scope) | No frontend test covers the expired → 400 → toast path (2026-07-13 failure-ledger precedent: frontend contract breaks need coverage) | **Tracked: #1397** |
| B1 | MEDIUM (pre-existing, fails safe, out of PR scope) | Third preview/apply asymmetry: Apply also runs `ValidatePermissionsAsync` (`AutomationExecutorService.cs:138`: board access 403, board existence 404, requester existence 404) — none checked at diff time; access revoked / board deleted mid-review previews 200 then applies 403/404 | **Tracked: #1398** |

### Summary
Core parity claims confirmed by both lenses. One in-PR LOW (comment drift) fixed; four out-of-scope findings tracked in #1397 and #1398 — zero silent drops. No merge-blocking findings remain from round 2.

**Comment by Chris0Jeky** [comment]
## Round 2 — Fixes Applied

| Finding | Severity | Disposition | Evidence |
|---------|----------|-------------|----------|
| F1: stale PaperReviewView.vue comments describing the pre-#1376 404 diff contract | LOW | Fixed in-PR (comment-only, no behavior change) | `b9545253` — verified with `npm run typecheck` (clean) + `npm run lint` (0 errors, 6 pre-existing warnings) |
| A1/A2/A3: expired/zero-op preview presentation in both review shells + missing frontend coverage for the 400 contract | MEDIUM/LOW (out of PR scope) | Tracked | #1397 |
| B1: third preview/apply asymmetry — `ValidatePermissionsAsync` access/existence gates not run at diff time | MEDIUM (out of PR scope, pre-existing, fails safe) | Tracked | #1398 |

All round-2 findings fixed or tracked; zero silent drops. Head is now `b9545253`.

File: backend/src/Taskdeck.Application/Services/AutomationProposalService.cs [hunk 1]
```diff
@@ -458,6 +475,18 @@ public async Task<Result<string>> GetProposalDiffAsync(Guid id, CancellationToke
...
+    /// Enforces the same expiry gate Apply runs via
+    /// <see cref="AutomationPolicyEngine.ValidatePolicy"/>: an expired proposal is rejected
+    /// with the identical <see cref="ErrorCodes.ValidationError"/> / "Proposal has expired"
+    /// shape. Diff callers run this after the structure gate (matching ValidatePolicy's order)
+    /// so preview rejects exactly what Apply would reject (#1376 preview == apply).
+    /// </summary>
+    private static Result ValidateProposalNotExpired(AutomationProposal proposal)
+        => DateTime.UtcNow > proposal.ExpiresAt
+            ? Result.Failure(ErrorCodes.ValidationError, "Proposal has expired")
+            : Result.Success();
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/AutomationProposalService.cs:488]: Fixed in 9f6eee1b — ValidateProposalNotExpired now delegates to the domain AutomationProposal.IsExpired property (one source of truth; identical semantics and error shape). Verified: AutomationProposalServiceTests 57 passed, Release build clean. Resolving.

File: backend/src/Taskdeck.Application/Services/AutomationProposalService.cs [hunk 2]
```diff
@@ -458,6 +475,18 @@ public async Task<Result<string>> GetProposalDiffAsync(Guid id, CancellationToke
...
+    /// <summary>
+    /// Enforces the same expiry gate Apply runs via
+    /// <see cref="AutomationPolicyEngine.ValidatePolicy"/>: an expired proposal is rejected
+    /// with the identical <see cref="ErrorCodes.ValidationError"/> / "Proposal has expired"
+    /// shape. Diff callers run this after the structure gate (matching ValidatePolicy's order)
+    /// so preview rejects exactly what Apply would reject (#1376 preview == apply).
+    /// </summary>
+    private static Result ValidateProposalNotExpired(AutomationProposal proposal)
+        => DateTime.UtcNow > proposal.ExpiresAt
+            ? Result.Failure(ErrorCodes.ValidationError, "Proposal has expired")
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/AutomationProposalService.cs:487]: Acknowledged and tracked in #1397 (finding 1 + addendum finding 4, with your terminal-status nuance recorded verbatim: the executor checks status first, so already-applied proposals short-circuit before ValidatePolicy, meaning preview==apply parity is not the operative contract for terminal reads — historical inspectability is). The status-agnostic expiry gate at diff time is the explicit direction of issue #1376 for pre-decision review flow (an expired proposal must not preview a diff Apply will reject); whether terminal/expired history stays inspectable — frontend stops calling /diff for terminal items and renders the stored preview with an EXPIRED/APPLIED banner, or the backend scopes the expiry gate by status — is flagged in #1397 as a product call for the maintainer. Not cycling that decision inside this PR. Resolving here; the thread is linked from the issue.

File: backend/src/Taskdeck.Application/Services/AutomationProposalService.cs [hunk 3]
```diff
@@ -435,10 +440,22 @@ public async Task<Result<string>> GetProposalDiffAsync(Guid id, CancellationToke
...
         // rejected at Apply cannot preview cleanly first (#1370 preview == apply). Apply runs
         // this via AutomationPolicyEngine.ValidatePolicy; mirror it here on the original path.
+        // A zero-operation proposal fails here with the same "Proposal must contain at least
+        // one operation" ValidationError Apply returns — previously this path returned the
+        // cached DiffPreview (200) or a 404, previewing a proposal Apply always rejects
+        // (#1376 preview == apply). Structure runs before expiry to match ValidatePolicy's
+        // order, so a proposal that is both empty and expired reports the empty error on both.
         var structureValidation = ProposalOperationStructureValidator.Validate(originalOperations);
         if (!structureValidation.IsSuccess)
             return Result.Failure<string>(structureValidation.ErrorCode, structureValidation.ErrorMessage);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/AutomationProposalService.cs:450]: Acknowledged and tracked in #1397 (addendum finding 5, citing this thread). Agreed the Paper no-op guard's local empty-preview bypass means the new backend structure rejection is never surfaced for zero-op-without-stored-preview proposals while the decision rail can still approve. This PR deliberately changes only the backend trust boundary (per issue #1376 scope; the guard comments were updated in b9545253 to describe the new 400 contract) — reconciling the Paper bypass and the approve affordance with the post-#1376 contract is the frontend slice specced in #1397. Resolving here; the thread is linked from the issue.
